Validate user-specified output bundles when processing coin selection requests

jonathanknowles opened the pull request:
# Issue Number

ADP-727
ADP-782

# Overview

This PR adds validation to coin selection requests, checking all **user-specified outputs** meet the following pre-conditions:

- output bundles do not include token quantities in excess of the maximum bound:
    `maxBound :: Word64`
- output bundles, when serialized, are within the size limit specified by the protocol:
    `4000` bytes
    
If either of the above pre-conditions are violated, an API error is returned, with a message that informs the user how they can solve the problem (usually by breaking up bundles into smaller bundles):

- `ErrOutputTokenBundleSizeExceedsLimit`
- `ErrOutputTokenQuantityExceedsLimit`

# Testing

This PR adds integration tests for both of the above scenarios:

- `WALLETS_COIN_SELECTION_07`
- `WALLETS_COIN_SELECTION_08`

# Refactoring

This PR also performs a small refactoring in the same spirit as the refactorings included in PR #2552.

We extract out `partition` functions for `Coin` and `TokenQuantity` values into the modules that define their types, similarly to the `X.equipartition` functions.

jonathanknowles committed:
Extract out coin partitioning into 'Coin.partition'.

This change makes coin partitioning follow the existing pattern, where
partitioning functions are located in the module that defines the type
on which they are operating.

jonathanknowles committed:
Extract out token quantity partitioning into 'TokenQuantity.partition'.

This change makes token quantity partitioning follow the existing
pattern, where partitioning functions are located in the module that
defines the type on which they are operating.

jonathanknowles committed:
Make `initSelectionCriteria` a top-level function.

jonathanknowles committed:
Make `initSelectionCriteria` reject excessive token quantities.

jonathanknowles committed:
Make `initSelectionCriteria` reject excessively large token bundles.

jonathanknowles committed:
Add integration test for `ErrTokenQuantityExceedsMaxBound`.

jonathanknowles committed:
Add integration test for `ErrTokenBundleSizeExceedsLimit`.

jonathanknowles committed:
Add error definitions to API specification.

Anviking commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/CoinSelections.hs:
```diff
@@ -293,6 +309,74 @@ spec = describe "SHELLEY_COIN_SELECTION" $ do
+-- Attempt to create a coin selection with an output that has an excessively
+-- high token quantity. (This should fail.)
+--
+wallets_coin_selection_07
```
Let's stick to the style of having the scenarios inline in the Spec?

jonathanknowles replied:
> Let's stick to the style of having the scenarios inline in the Spec?

I did it this way because including the scenarios inline involved a great deal of excessive wrapping (12 columns are already eaten up by the indentation).

But I don't feel strongly about it: I can re-inline them if you feel strongly about it. 👍🏻

jonathanknowles replied:
Fixed in 62944743aafa344b1b09935da53f900436818c13.

Anviking commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -2866,8 +2869,45 @@ instance LiftHandler (ErrInvalidDerivationIndex 'Soft level) where
+    handler e = apiError err403 TokenQuantityExceedsMaxBound $ mconcat
+        [ "One of the asset quantities you've specified is greater than the "
+        , "maximum quantity allowed in a single transaction output. Try "
+        , "splitting this quantity across two or more outputs. "
```
This is because the cap for a single output is `maxBound :: Word64`, but quantities in general are `Natural`?

Ah, yes.

jonathanknowles replied:
That's correct. But the cap only applies to single outputs.

It's quite possible to get **_more than_** `maxBound :: Word64` for a given asset if we're summing across a whole UTxO set to find a wallet's balance.

Anviking commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -2866,8 +2869,45 @@ instance LiftHandler (ErrInvalidDerivationIndex 'Soft level) where
+        , ". Token quantity specified: "
+        , pretty (view #quantity e)
+        , ". Maximum allowable token quantity: "
+        , pretty (view #quantityMaxBound e)
```
Nice!

jonathanknowles committed:
Inline integration test definitions.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2560#discussion_r593012784

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs:
```diff
@@ -1233,19 +1229,7 @@ makeChangeForCoin
-  where
-    zeroWeightSum :: HasCallStack => a
-    zeroWeightSum = error
-        "partitionValue: The specified weights must have a non-zero sum."
```
Before this PR, this error message actually had the wrong function name. (Now fixed.)

jonathanknowles commented on lib/core/src/Cardano/Wallet/Transaction.hs:
```diff
@@ -168,6 +175,37 @@ defaultTransactionCtx = TransactionCtx
+    { address :: !Address
+      -- ^ The address to which this token bundle was to be sent.
+    , assetCount :: !Int
+      -- ^ The number of assets within the token bundle.
```
Note that it might be possible to be more helpful here, by giving a suggestion of **_how many bundles_** the user will need to split their oversized bundle into.

But this is harder than it sounds, as different partitionings will give rise to different serialized lengths. Consider this example:

- Bundle **A**: ❌ (_too large_)
    - `100` assets with names of length `1`.
    - `100` assets with names of length `32`.

We might determine that it's feasible to split this into 2 valid bundles, if the bundles are partitioned this way:

- Bundle **A1**: 🆗 
    - `50` assets with names of length `1`.
    - `50` assets with names of length `32`.
- Bundle **A2**: 🆗 
    - `50` assets with names of length `1`.
    - `50` assets with names of length `32`.

But if the user is not familiar with how bundles are serialized, they might end up partitioning the bundles like this:

- Bundle **A3**: 🆗 
    - `100` assets with names of length `1`.
- Bundle **A4**: ❌ (_too large_)
    - `100` assets with names of length `32`.

In the above example, bundle **A4** is still likely to be rejected.

jonathanknowles replied:
Another option (for the future) would be to provide a suggested partitioning in the API error response, or to provide automated splitting of output bundles (should the user request this).

Anviking commented on specifications/api/swagger.yaml:
```diff
@@ -2693,6 +2693,33 @@ x-errInvalidCoinSelection: &errInvalidCoinSelection
+        quantity that exceeds the maximum bound supported by the protocol.
+    code:
+      type: string
+      enum: ['token_quantity_exceeds_max_bound']
```
How about calling the errors `too_many_tokens[_in_output]` and `too_large_token`?

It takes me a while to parse the difference between the the current terms...

token_bundle_size_exceeds_limit
token_quantity_exceeds_max_bound

They both start with `token`, end with `exceeds limit ≈ max bound`.

Although —`too_large_token` — maybe tokens can't have largeness 🤔

jonathanknowles replied:
@Anviking thanks for this comment. I was also a bit unhappy with the naming.

I agree it's very hard to come up with names that are completely unambiguous, since `camelCase` and `snake_case` make it difficult to see the groupings. And using clarification words like "of" and "the" would probably make the names longer than we'd like.

The other confusing aspect is the distinction between **_asset_** and **_token_**.

Just to share with you my thinking: I tend to think of a "token quantity" as a **_discrete quantity_** of some **_asset_**, in the same way we've historically thought of a "coin quantity" as being a **_discrete quantity_** of ada (the asset).

I also think that:

- since these errors pertain to **_outputs_**, the name `Output` should probably appear in the name of each error.
- since both errors pertain to **_exceeding some limit_**, it's better to use consistent terminology. I've gone for "exceeds limit" in both cases.

I've revised the names to the following:

| Name | Meaning |
| -- | -- |
| `output_token_quantity_exceeds_limit` | A user-specified transaction **_output_** contains, for some asset, a **_token quantity_** that **_exceeds_** the **_limit_** supported by the protocol. |
| `output_token_bundle_size_exceeds_limit` | A user-specified transaction **_output_** contains a **_token bundle_** whose **_size_** (serialized length) **_exceeds_** the **_limit_** supported by the protocol. |

See: 7322723cbb9d7d44e020a8f8fb79fe62889da3ae

jonathanknowles committed:
Revise error terminology.

We now have:

  - output_token_bundle_size_exceeds_limit:
    The size of an output's token bundle exceeds the limit.

  - output_token_quantity_exceeds_limit:
    A token quantity within an output exceeds the limit.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2560#discussion_r593334380

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -432,6 +424,82 @@ _estimateMaxNumberOfInputs txMaxSize ctx outs =
+        Left $
+            -- We encountered one or more excessive token quantities.
+            -- Just report the first such quantity:
+            ErrSelectionCriteriaOutputTokenQuantityExceedsLimit $
```
Just wondering: did you consider adding this logic to `performSelection` instead?

jonathanknowles replied:
> Just wondering: did you consider adding this logic to `performSelection` instead?

That's a good point. The `performSelection` function would also be a perfectly valid place to put this validation logic.

However, doing so would mean a little more work, as we'd have to adjust the property tests for `performSelection`. It might not be too hard though.

Anviking reviewed (approved):
lgtm!
